enhancements/api-review/custom-tags-aws.md

```diff
@@ -0,0 +1,143 @@
+will apply these tags to all AWS resources they create.
+
+userTags that are specified in the infrastructure resource will merge with userTags specified in an individual component. In the case where a userTag is specified in the infrastructure resource and there is a tag with the same key specified in an individual component, the value from the individual component will have precedence and be used.
+
+The userTags field is intended to be set at install time and is considered immutable. Components that respect this field must only ever add tags that they retrieve from this field to cloud resources, they must never remove tags from the existing underlying cloud resource even if the tags are removed from this field(despite it being immutable).
+
+If the userTags field is changed post-install, there is no guarantee about how an in-cluster operator will respond to the change. Some operators may reconcile the change and change tags on the AWS resource. Some operators may ignore the change. However, if tags are removed from userTags, the tag will not be removed from the AWS resource.
+
+### User Stories
+
+https://issues.redhat.com/browse/SDE-1146 - IAM users and roles can only operate on resources with specific tags
+As a security-conscious ROSA customer, I want to restrict the permissions granted to Red Hat in my AWS account by using
+AWS resource tags.
+
+### Risks and Mitigations
+
+## Design Details
+
+### Test Plan
+
+TBD
+
+### Graduation Criteria
+
+TBD
+
+#### Dev Preview -> Tech Preview
+
+TBD
+
+#### Tech Preview -> GA
+
+TBD
+
+#### Removing a deprecated feature
+
+This enhancement does not remove or deprecate any features.
+
+### Upgrade / Downgrade Strategy
+
+On upgrade:
+
+The new status field won't be populated since it is only populated by the installer and that can't have happened if the cluster was installed from a prior version. Components that consume the new field should take no action since they will see no additional tags.
+
+On downgrade:
+
+The status field may remain populated, components may or may not continue to tag newly created resources w/ the additional tags depending on whether or not a given component still has logic to respect the status tags, after the downgrade.
+
+### Version Skew Strategy
+
+TBD
+
+## Implementation History
+
+## Drawbacks
+
+If a customer decides they do not want these tags applied to new resources, there is no clean way to address that need:
+
+1. they would have to get help from support to edit the status field
+2. they would have to manually remove the undesired tags from any existing resources
+
+In the future we will want to introduce a spec field where a customer can specify and edit these tags, which will be reflected into status when changed, and we will expect consuming components to reconcile changes to the set of tags by applying net new tags to existing resources (but they will still not remove tags that are dropped from the list of tags).
+
+## Alternatives
+
+TBD
```

Comment:
Not sure what other options have already been discussed, but there are several existing auto-tagging projects in the wild.

They generally work by:

- Watching CloudTrail for events
- Adding tags to resources in the events based on some criteria

A simple policy would be something like "tag everything created by X" (and then the installer and all on-cluster components would be sure to use X for AWS interactions)

@estroz and I hacked on something that did this here https://github.com/coreos/grafiti, but I'm sure there are other, newer projects that do something similar.

Reply:
The main requirement for security-minded customers is that tags be applied to resources atomically at creation-time. Customers set up AWS access such that any attempt to create an AWS resource without a tag like "red-hat-managed: true" in the definition is denied. Looks like grafiti scans and tags post-creation, which won't work for this use case.

Reply:
It would be good to capture those rejected alternatives, and the reasons, in the body of the doc (either in this draft or an update).

Reply:
I think it is?
`tags must be applied at creation time, in an atomic operation. It isn't acceptable to create an object and, after some period of time, apply the tags post-creation.`

Reply:
Sorry, I meant listing the other tool that @ecordell mentioned here in the alternatives section and pointing out that it does not meet the requirements described above.

The idea, at least for me, of having the alternatives section is to capture these sorts of conversations, especially why other approach will _not_ work. That sort of detail is useful to future maintainers.

Reply:
> Customers set up AWS access such that any attempt to create an AWS resource without a tag like "red-hat-managed: true" in the definition is denied

Do you have any links I could read to learn how this is done?

Reply:
@ecordell checkout https://docs.aws.amazon.com/IAM/latest/UserGuide/access_tags.html and https://docs.aws.amazon.com/IAM/latest/UserGuide/tutorial_attribute-based-access-control.html

Even though it's actually not very common to do it, most security-minded companies would separate AWS accounts and run their workloads using a different IAM user so you can distinguish later on in CloudTrail what resource originated from where.

> AWS access such that any attempt to create an AWS resource without a tag like "red-hat-managed: true" in the definition is denied. Looks like grafiti scans and tags post-creation, which won't work for this use case.

Not to derail that thread too much, but wouldn't we want introduce such a functionality ourselves first to ensure we always tag these resources? I can see some projects down the road not knowing about the tags and creating resources - would be great to catch that during CI if possible.